yewtil/src/not_equal_assign.rs

```diff
@@ -43,13 +43,74 @@ pub trait NeqAssign<NEW> {
     ///#         unimplemented!()
     ///#     }
     ///  }
+    ///
+    /// let mut foo = 1;
+    ///
+    /// assert_eq!(foo.neq_assign(42), true);
+    /// assert_eq!(foo, 42);
+    ///
+    /// assert_eq!(foo.neq_assign(42), false);
     /// ```
     fn neq_assign(&mut self, new: NEW) -> ShouldRender;
 }
 
 impl<T: BorrowMut<U>, U: PartialEq> NeqAssign<U> for T {
     fn neq_assign(&mut self, new: U) -> bool {
-        if self.borrow() != &new {
+        self.neq_assign_by(new, |x, y| x == y)
+    }
+}
+/// Blanket trait to provide a convenience method for assigning props in `changed` or updating values in `update`.
+///
+/// Like `neq_assign`, but for cases where `self` doesn't impl `PartialEq` or a nonstandard equality comparison is needed.
+///
+/// Useful for `Result<T, E: !PartialEq>`.
+pub trait NeqAssignBy<NEW> {
+    /// ```
+    /// # use yewtil::{NeqAssign, NeqAssignBy};
+    /// ##[derive(Clone, Debug)]
+    /// struct NonComparableError;
+    ///
+    /// fn eq_by_ok<T, E>(a: &Result<T, E>, b: &Result<T, E>) -> bool
+    /// where
+    ///     T: PartialEq,
+    /// {
+    ///     match (a, b) {
+    ///         (Ok(_), Err(_))
+    ///         | (Err(_), Ok(_))
+    ///         | (Err(_), Err(_)) => false,
+    ///         (Ok(a), Ok(b)) => a == b,
+    ///     }
+    /// }
+    ///
+    /// let mut foo: Result<u32, NonComparableError> = Ok(1);
+    ///
+    /// // Won't compile
+    /// // assert_eq!(foo.neq_assign(Ok(42)), true)
+    ///
+    /// assert_eq!(foo.neq_assign_by(Ok(42), eq_by_ok), true);
+    /// assert_eq!(foo.clone().unwrap(), 42);
+    ///
+    /// assert_eq!(foo.neq_assign_by(Err(NonComparableError), eq_by_ok), true);
+    /// assert!(foo.is_err());
+    ///
+    /// // The tradeoff: all assignments of an `Err` value will count as updates, even if they are
+    /// // "the same" for all practical intents and purposes.
+    /// assert_eq!(foo.neq_assign_by(Err(NonComparableError), eq_by_ok), true);
+    /// assert_eq!(foo.neq_assign_by(Err(NonComparableError), eq_by_ok), true);
+    /// assert_eq!(foo.neq_assign_by(Err(NonComparableError), eq_by_ok), true);
+    /// ```
+    ///
+    fn neq_assign_by<F>(&mut self, new: NEW, eq: F) -> ShouldRender
+    where
+        F: FnOnce(&NEW, &NEW) -> bool;
+}
+
+impl<T: BorrowMut<U>, U> NeqAssignBy<U> for T {
+    fn neq_assign_by<F>(&mut self, new: U, eq: F) -> ShouldRender
+    where
+        F: FnOnce(&U, &U) -> bool,
+    {
+        if !eq(self.borrow(), &new) {
```

Comment:
Try `if ne(self.borrow(), &new) {...` instead

Reply:
`!=` would rely on `PartialEq` just like the original `NeqAssign` trait does.

I chose `eq` instead of `ne` for the passed closure for parallelism with `std`. In the case of existing comparison functions (instead of writing an ad-hoc closure), standard practice seems to be checking for equality rather than for inequality. At least in `std`, various `ne`s are implemented in terms of their analogous `eq`s.

To me, the semantics of `*_by` hints "supply an alternate implementation of the trait behavior". Since in this case that's `PartialEq`, my first thought would be "oh, pass an alternate `eq` function". That would be the Principle of Least Surprise for me, but I could see how it might be interpreted differently.

Reply:
Yeah, this comment was plainly wrong on my part. Sorry for making you justify the naming decision, it totally makes sense as it is now.

Reply:
No problem, it's always good to verify my intuitions!